File: sdk/typescript/test/normalize.ts
// Copyright 2026 AI5Labs Research OPC Private Limited
// SPDX-License-Identifier: Apache-2.0

/**
 * TypeScript port of `sdk/python/tests/conformance/normalize.py`.
 *
 * Produces the same stable, comparable shape from a captured span so the
 * TS-emitted telemetry can be deep-equal-asserted against the SAME golden
 * JSON files the Python conformance suite uses. Matches the Python logic:
 *
 * - drops trace/span/parent ids, timestamps, duration
 * - drops the OTel `exception` event
 * - sorts attribute keys
 * - normalizes UUID-/latency-bearing attribute values to a placeholder
 * - KEEPS sha-256 hashes (part of the contract)
 * - orders spans by name
 */

import type { ReadableSpan } from "@opentelemetry/sdk-trace-node";
import { SpanKind, SpanStatusCode } from "@opentelemetry/api";

const PLACEHOLDER = "<normalized>";

const UUID_ATTR_KEYS = new Set<string>([
  "fabric.checkpoint.checkpoint_id",
  "fabric.eval.eval_id",
  "fabric.policy.evaluation_id",
  "fabric.judge.request_id",
]);

const LATENCY_ATTR_KEYS = new Set<string>([
  "fabric.guardrail.latency_ms",
  "fabric.policy.latency_ms",
  "fabric.retrieval.latency_ms",
]);

type Json = unknown;

function normalizeAttrValue(value: unknown): Json {
  if (Array.isArray(value)) {
    return value.map((v) => normalizeAttrValue(v));
  }
  return value as Json;
}

export function normalizeAttributes(
  attributes: Record<string, unknown> | undefined,
): Record<string, Json> {
  const out: Record<string, Json> = {};
  if (!attributes) {
    return out;
  }
  for (const key of Object.keys(attributes).sort()) {
    if (UUID_ATTR_KEYS.has(key) || LATENCY_ATTR_KEYS.has(key)) {
      out[key] = PLACEHOLDER;
    } else {
      out[key] = normalizeAttrValue(attributes[key]);
    }
  }
  return out;
}

// OTel JS exposes SpanKind as a numeric enum; the goldens use the Python
// SpanKind *names* (INTERNAL/CLIENT/...). Map to the same strings.
const SPAN_KIND_NAME: Record<number, string> = {
  [SpanKind.INTERNAL]: "INTERNAL",
  [SpanKind.SERVER]: "SERVER",
  [SpanKind.CLIENT]: "CLIENT",
  [SpanKind.PRODUCER]: "PRODUCER",
  [SpanKind.CONSUMER]: "CONSUMER",
};

// Likewise StatusCode: OTel JS uses UNSET/OK/ERROR numeric enum; Python
// goldens carry the names. Map to UNSET/OK/ERROR strings.
const STATUS_CODE_NAME: Record<number, string> = {
  [SpanStatusCode.UNSET]: "UNSET",
  [SpanStatusCode.OK]: "OK",
  [SpanStatusCode.ERROR]: "ERROR",
};

interface NormalizedEvent {
  name: string;
  attributes: Record<string, Json>;
}

interface NormalizedSpan {
  name: string;
  kind: string;
  status: { code: string; description: string | null };
  attributes: Record<string, Json>;
  events: NormalizedEvent[];
}

function normalizeEvent(event: {
  name: string;
  attributes?: Record<string, unknown>;
}): NormalizedEvent {
  return {
    name: event.name,
    attributes: normalizeAttributes(event.attributes),
  };
}

export function normalizeSpan(span: ReadableSpan): NormalizedSpan {
  // OTel JS carries the status description on `message` and leaves it
  // undefined when unset; Python emits the status description as null.
  // Normalize undefined/"" -> null.
  const description = span.status.message ? span.status.message : null;
  const events = span.events
    .filter((e) => e.name !== "exception")
    .map((e) => normalizeEvent({ name: e.name, attributes: e.attributes }));
  return {
    name: span.name,
    kind: SPAN_KIND_NAME[span.kind] ?? String(span.kind),
    status: {
      code: STATUS_CODE_NAME[span.status.code] ?? String(span.status.code),
      description,
    },
    attributes: normalizeAttributes(span.attributes),
    events,
  };
}

export function normalizeSpans(spans: ReadableSpan[]): NormalizedSpan[] {
  const normalized = spans.map((s) => normalizeSpan(s));
  normalized.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  return normalized;
}

File: sdk/python/tests/conformance/normalize.py
# Copyright 2026 AI5Labs Research OPC Private Limited
# SPDX-License-Identifier: Apache-2.0
"""Normalization of captured spans into stable, comparable dicts.

The goal is a representation that is byte-identical across runs and
machines for a fixed scenario input. We therefore drop or zero the
non-deterministic fields (ids, timestamps, durations, latencies, and
generated UUIDs) while keeping everything that is part of the wire
contract — span name, kind, status, the ``fabric.*`` / ``gen_ai.*``
attribute keys + values, and the ordered list of span events.

Hashes are deterministic for a fixed input and are therefore KEPT:
they are part of the contract a downstream consumer must reproduce.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence

    from opentelemetry.sdk.trace import ReadableSpan
    from opentelemetry.util.types import AttributeValue

# Placeholder substituted for any normalized-away volatile value.
PLACEHOLDER = "<normalized>"

# Attribute keys whose values are generated UUIDs. Stable across the
# contract by key, volatile by value — replace with the placeholder.
_UUID_ATTR_KEYS = frozenset(
    {
        "fabric.decision_id",
        "fabric.checkpoint.checkpoint_id",
        "fabric.eval.eval_id",
        "fabric.policy.evaluation_id",
        "fabric.judge.request_id",
    }
)

# Attribute keys carrying wall-clock-dependent latencies. The key is
# part of the contract; the value is not reproducible.
_LATENCY_ATTR_KEYS = frozenset(
    {
        "fabric.guardrail.latency_ms",
        "fabric.policy.latency_ms",
        "fabric.retrieval.latency_ms",
    }
)


def _normalize_attr_value(value: AttributeValue) -> Any:
    """Convert an OTel attribute value to a JSON-stable form.

    Tuples/lists (OTel sequence attributes) become lists; scalars pass
    through unchanged.
    """
    if isinstance(value, (tuple, list)):
        return [_normalize_attr_value(v) for v in value]
    return value


def normalize_attributes(attributes: Mapping[str, AttributeValue] | None) -> dict[str, Any]:
    """Return a sorted-by-key dict with volatile values placeheld.

    UUID-bearing and latency-bearing keys keep their key (so the
    contract still asserts the key is present) but get the stable
    placeholder value.
    """
    if not attributes:
        return {}
    out: dict[str, Any] = {}
    for key in sorted(attributes):
        if key in _UUID_ATTR_KEYS or key in _LATENCY_ATTR_KEYS:
            out[key] = PLACEHOLDER
        else:
            out[key] = _normalize_attr_value(attributes[key])
    return out


def _normalize_event(event: Any) -> dict[str, Any]:
    """Normalize one span event into ``{name, attributes}``.

    The event timestamp is dropped entirely (ordering is preserved by
    list position, which is the contract-relevant signal).
    """
    return {
        "name": event.name,
        "attributes": normalize_attributes(event.attributes),
    }


def normalize_span(span: ReadableSpan) -> dict[str, Any]:
    """Normalize one span into a stable, comparable dict.

    Dropped: trace_id, span_id, parent id, start/end timestamps,
    duration. Kept (normalized): name, kind, status code + description,
    attributes, ordered ``fabric.*`` / ``gen_ai.*`` events.

    The OTel-internal ``exception`` event (emitted by
    ``span.record_exception`` on fail-closed paths) is dropped: its
    ``exception.stacktrace`` carries machine-dependent file paths and
    line numbers and is not part of the Fabric wire contract.
    """
    status = span.status
    events = [_normalize_event(e) for e in span.events if e.name != "exception"]
    return {
        "name": span.name,
        "kind": span.kind.name,
        "status": {
            "code": status.status_code.name,
            "description": status.description,
        },
        "attributes": normalize_attributes(span.attributes),
        "events": events,
    }


def normalize_spans(spans: Sequence[ReadableSpan]) -> list[dict[str, Any]]:
    """Normalize a captured span list, ordering deterministically.

    Spans are emitted in finish order, which for a single decision is
    children-before-parent. To make the golden independent of timing we
    sort: child spans (``fabric.llm_call`` / ``fabric.tool_call``)
    first by name, then the parent ``fabric.decision``. Within a
    scenario each span name is unique, so name ordering is total.
    """
    normalized = [normalize_span(s) for s in spans]
    normalized.sort(key=lambda s: s["name"])
    return normalized
